carta/session.py

```diff
@@ -0,0 +1,619 @@
+Alternatively, the user can create a new session which runs in a headless browser controlled by the wrapper. The user can connect to an existing CARTA backend instance (using the :obj:`carta.session.Session.connect` classmethod), or first start a new CARTA backend instance which is also controlled by the wrapper (using the :obj:`carta.session.Session.new` classmethod). The backend can be started either on the local host or on a remote host which the user can access with passwordless SSH.
+"""
+
+import base64
+
+from .image import Image
+from .constants import CoordinateSystem, LabelType, BeamType, PaletteColor, Overlay
+from .protocol import Protocol
+from .util import logger, Macro, split_action_path, CartaScriptingException
+from .validation import validate, String, Number, Color, Constant, Boolean, NoneOr, OneOf
+
+
+class Session:
+    """This object corresponds to a CARTA frontend session.
+
+    This class provides the core generic method for calling actions on the frontend (through the backend), as well as convenience methods which wrap this generic method and provide a more intuitive and user-friendly interface to frontend functionality associated with the session as a whole.
+
+    This class should not be instantiated directly. Three class methods are provided for creating different types of sessions with all the appropriate parameters set: :obj:`carta.session.Session.interact` for interacting with an existing CARTA session open in the user's browser, :obj:`carta.session.Session.create` for creating a new CARTA session in a headless browser by connecting to an existing CARTA backend or controller instance, and :obj:`carta.session.Session.start_and_create` for starting a new backend instance and then connecting to it to create a new session in a headless browser.
+
+    The session object can be used to create image objects, which provide analogous convenience methods for functionality associated with individual images.
+
+    Parameters
+    ----------
+    session_id : integer
+        The ID of the CARTA frontend session associated with this object. This is set automatically when a new session is created with :obj:`carta.session.Session.create` or :obj:`carta.session.Session.start_and_create`.
+    protocol : :obj:`carta.protocol.Protocol`
+        The protocol object which handles HTTP communication with the CARTA scripting API. This is created automatically when a new session is created with one of the three class methods for creating sessions.
+    browser : :obj:`carta.browser.Browser`
+        The browser object associated with this session. This is created automatically when a new session is created with :obj:`carta.session.Session.create` or :obj:`carta.session.Session.start_and_create`.
+    backend : :obj:`carta.backend.Backend`
+        The backend object associated with this session. This is created automatically when a new session is created with :obj:`carta.session.Session.start_and_create`.
+    debug_no_auth : boolean
```

Comment:
Ah, that's left over from when it was a parameter to this class. I will remove it.

I haven't been very consistent about listing attributes (since they're often the same as the parameters); I'll go through the classes and add them in. I believe that the convention is to list all public attributes (so not the ones with leading underscores which are intended to be private).

Reply:
I see. maybe we can shorten the descriptions when they're the same as the parameters